ocaml/gencert/validation.ml

```diff
@@ -0,0 +1,247 @@
+      (Certificate.issuer cert))
+
+let ext_authority_matches_subject trusted cert =
+  match
+    X509.(
+      Extension.
+        ( find Authority_key_id (Certificate.extensions cert)
+        , find Subject_key_id (Certificate.extensions trusted) ))
+  with
+  | _, None | None, _ ->
+      (* not mandatory *)
+      true
+  | Some (_, (Some auth, _, _)), Some (_, au) ->
+      Cstruct.equal auth au
+  (* TODO: check exact rules in RFC5280 *)
+  | Some (_, (None, _, _)), _ ->
+      (* not mandatory *)
+      true
+
+let signs pathlen trusted cert =
+  match
+    ( issuer_matches_subject trusted cert
+    , ext_authority_matches_subject trusted cert
+    , X509.Validation.validate_signature supported_cert_hashes trusted cert
+    , X509.Validation.validate_path_len pathlen trusted )
+  with
+  | true, true, Ok (), true ->
+      Ok ()
+  | _ ->
+      Error (`Msg (server_certificate_chain_invalid, []))
+
+let rec validate_anchors pathlen cert = function
+  | [] ->
+      Error (`Msg (server_certificate_chain_invalid, []))
+  | x :: xs -> (
+    match signs pathlen x cert with
+    | Ok _ ->
+        Ok x
+    | Error _ ->
+        validate_anchors pathlen cert xs
+  )
+
+let issuers_of trusted cert =
+  List.filter (fun p -> issuer_matches_subject p cert) trusted
+
+let verify_single_chain time anchors chain =
+  let open X509.Validation in
+  let rec climb pathlen = function
+    | cert :: issuer :: certs ->
+        validate_intermediate_certificate issuer time >>= fun _ ->
+        signs pathlen issuer cert >>= fun () ->
+        climb (succ pathlen) (issuer :: certs)
+    | [c] ->
+        let anchors = issuers_of anchors c in
+        validate_anchors pathlen c anchors
+    | [] ->
+        Error (`Msg (server_certificate_chain_invalid, []))
+  in
+  climb 0 chain
+
+let valid_chains time anchors chains =
+  let rec loop acc chains =
+    match (acc, chains) with
+    | [], [] ->
+        Error (`Msg (server_certificate_chain_invalid, []))
+    | _, [] ->
+        Ok acc
+    | _, chain :: chains -> (
+      match verify_single_chain time anchors chain with
+      | Ok root ->
+          loop (root :: acc) chains
+      | Error _ ->
+          loop acc chains
+    )
+  in
+  loop [] chains
+
+let verify_chain_of_trust ~server ~private_key ~intermediates ~time ~anchors =
+  validate_server_certificate server time private_key >>= fun server ->
+  (* build all paths *)
+  let paths = X509.Validation.build_paths server intermediates in
+  let anchors = List.filter (has_valid_time time) anchors in
+  (* return valid chains *)
+  valid_chains time anchors paths
+
+let load_ca_cert_bundle ca_cert_bundle_path =
+  Xapi_stdext_unix.Unixext.string_of_file ca_cert_bundle_path
+  |> Cstruct.of_string
+  |> X509.Certificate.decode_pem_multiple
+  |> function
+  | Ok (_ :: _ as trust_anchors) ->
+      Ok trust_anchors
+  | _ ->
+      Error (`Msg (server_certificate_chain_invalid, []))
+
+let validate_certificate_chain pem_chain =
```

Comment:
Is this a good name? The function decodes the PEM and does some syntactic checking but is this a validation? The name implies more to me.

Reply:
This the "validation" that is done to chains in 8.2. I agree it doesn't provide much value right now, but I'm unsure on how to factor it out. It's a consequence on the function signature chosen for `install_server_certificate`.

Reply:
This has been renamed to `decode_certificate_chain`.